File: android-app/app/src/main/java/lvt/crm/data/auth/AuthRepository.kt
package lvt.crm.data.auth

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.concurrent.atomic.AtomicLong
import lvt.crm.data.convex.ConvexException
import lvt.crm.data.convex.AuthApi
import lvt.crm.data.convex.ConvexHttpClient
import org.json.JSONObject

interface SignInGateway {
    suspend fun signIn(email: String, password: String): Result<UserSession>
    suspend fun requestPasswordReset(email: String): Result<Unit>
}

class AuthRepository(
    private val tokenStore: CredentialStore,
    private val convex: AuthApi,
    private val beforeSignOut: (suspend (String) -> Unit)? = null,
    private val afterAuthenticated: (suspend () -> Result<*>)? = null,
) : SignInGateway {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val _state = MutableStateFlow<AuthState>(AuthState.Loading)
    val state: StateFlow<AuthState> = _state.asStateFlow()
    private val authGeneration = AtomicLong(0)
    private val credentialTransitionLock = Any()

    init {
        scope.launch { restoreSession() }
    }

    suspend fun restoreSession() {
        val generation = authGeneration.get()
        val credentials = tokenStore.snapshot()
        if (credentials == null) {
            synchronized(credentialTransitionLock) {
                if (authGeneration.get() == generation) _state.value = AuthState.SignedOut
            }
            return
        }
        try {
            val session = fetchSession()
            if (session == null) {
                clearCurrentCredentialsAndSignOut(credentials, generation)
            } else {
                publish(session, generation, credentials)
                runCatching { afterAuthenticated?.invoke() }
            }
        } catch (e: Exception) {
            // Keep tokens if network blip; only clear on hard auth failure.
            if (e is ConvexException && (
                    e.code.contains("Unauthenticated", ignoreCase = true) ||
                        e.code.contains("Authentication", ignoreCase = true)
                    )
            ) {
                clearCurrentCredentialsAndSignOut(credentials, generation)
            } else {
                synchronized(credentialTransitionLock) {
                    if (
                        authGeneration.get() == generation &&
                        tokenStore.snapshot() == credentials
                    ) {
                        // Leave the loading state so the user can retry login.
                        _state.value = AuthState.SignedOut
                    }
                }
            }
        }
    }

    override suspend fun signIn(email: String, password: String): Result<UserSession> {
        val normalized = email.trim().lowercase()
        if (normalized.isEmpty() || password.isEmpty()) {
            return Result.failure(IllegalArgumentException("EMAIL_PASSWORD_REQUIRED"))
        }
        val (previousCredentials, previousState, generation, credentialRevision) =
            synchronized(credentialTransitionLock) {
                val credentials = tokenStore.snapshot()
                val state = _state.value
                val nextGeneration = authGeneration.incrementAndGet()
                val nextRevision = tokenStore.invalidatePendingWrites()
                if (credentials != null && state.isAuthenticated()) {
                    _state.value = AuthState.Loading
                }
                SignInAttempt(credentials, state, nextGeneration, nextRevision)
            }
        var credentialsIssued = false
        var issuedCredentials: CredentialSnapshot? = null
        return try {
            val args = JSONObject()
                .put("provider", "password")
                .put(
                    "params",
                    JSONObject()
                        .put("email", normalized)
                        .put("password", password)
                        .put("flow", "signIn"),
                )
            val result = convex.action("auth:signIn", args, authenticated = false)
            val tokens = result.optJSONObject("tokens")
                ?: throw ConvexException("NO_TOKENS", "Đăng nhập không trả về token.")
            val access = tokens.getString("token")
            val refresh = tokens.getString("refreshToken")
            credentialsIssued = synchronized(credentialTransitionLock) {
                authGeneration.get() == generation &&
                    tokenStore.saveIfRevision(credentialRevision, access, refresh)
            }
            if (!credentialsIssued) {
                return Result.failure(ConvexException("SIGN_IN_SUPERSEDED"))
            }
            issuedCredentials = CredentialSnapshot(access, refresh, credentialRevision)

            val session = fetchSession()
                ?: throw ConvexException("NO_SESSION", "Không tải được phiên đăng nhập.")
            if (!publish(session, generation)) {
                return Result.failure(ConvexException("SIGN_IN_SUPERSEDED"))
            }
            runCatching { afterAuthenticated?.invoke() }
            Result.success(session)
        } catch (e: ConvexException) {
            if (previousCredentials != null && previousState.isAuthenticated()) {
                rollbackAccountSwitch(
                    generation,
                    credentialRevision,
                    issuedCredentials,
                    previousCredentials,
                    previousState,
                )
            } else if (credentialsIssued && e.invalidatesCredentials()) {
                issuedCredentials?.let { clearCurrentCredentialsAndSignOut(it, generation) }
            }
            Result.failure(e)
        } catch (e: Exception) {
            if (previousCredentials != null && previousState.isAuthenticated()) {
                rollbackAccountSwitch(
                    generation,
                    credentialRevision,
                    issuedCredentials,
                    previousCredentials,
                    previousState,
                )
            }
            Result.failure(ConvexException("SIGN_IN_FAILED", e.message ?: "SIGN_IN_FAILED"))
        }
    }

    override suspend fun requestPasswordReset(email: String): Result<Unit> {
        val normalized = email.trim().lowercase()
        if (normalized.isEmpty()) {
            return Result.failure(IllegalArgumentException("EMAIL_REQUIRED"))
        }
        return try {
            convex.action(
                "users:requestPasswordReset",
                JSONObject().put("email", normalized),
                authenticated = false,
            )
            Result.success(Unit)
        } catch (e: ConvexException) {
            Result.failure(e)
        } catch (e: Exception) {
            Result.failure(ConvexException("PASSWORD_RESET_FAILED", e.message ?: "PASSWORD_RESET_FAILED"))
        }
    }

    suspend fun changePassword(newPassword: String): Result<Unit> {
        val generation = authGeneration.get()
        if (newPassword.length < 8) {
            return Result.failure(ConvexException("PASSWORD_TOO_SHORT", ConvexHttpClient.humanize("PASSWORD_TOO_SHORT")))
        }
        try {
            convex.action("users:changeOwnPassword", JSONObject().put("newPassword", newPassword))
        } catch (e: ConvexException) {
            return Result.failure(e)
        } catch (e: Exception) {
            return Result.failure(ConvexException("PASSWORD_CHANGE_FAILED", e.message ?: "PASSWORD_CHANGE_FAILED"))
        }
        return try {
            val session = fetchSession()
            if (session != null) {
                publish(session, generation)
            } else if (authGeneration.get() == generation) {
                // Password changed; force re-login if session vanished
                synchronized(credentialTransitionLock) {
                    if (authGeneration.get() == generation) {
                        authGeneration.incrementAndGet()
                        tokenStore.clear()
                        _state.value = AuthState.SignedOut
                    }
                }
            }
            Result.success(Unit)
        } catch (e: Exception) {
            // The mutation is the commit point. A refresh failure must not be reported as a failed change.
            Result.success(Unit)
        }
    }

    fun signOut() {
        val accessToken = synchronized(credentialTransitionLock) {
            val captured = tokenStore.accessToken
            authGeneration.incrementAndGet()
            tokenStore.clear()
            _state.value = AuthState.SignedOut
            captured
        }
        if (accessToken.isNullOrBlank()) return
        scope.launch {
            try {
                beforeSignOut?.invoke(accessToken)
            } catch (_: Exception) {
                // Device cleanup is independent from server-side token revocation.
            }
            try {
                convex.actionWithToken("auth:signOut", JSONObject(), accessToken)
            } catch (_: Exception) {
                // Local credentials are already revoked; remote cleanup is best-effort.
            }
        }
    }

    private fun publish(
        session: UserSession,
        generation: Long = authGeneration.get(),
        expectedCredentials: CredentialSnapshot? = null,
    ): Boolean =
        synchronized(credentialTransitionLock) {
            if (
                authGeneration.get() != generation ||
                expectedCredentials != null && tokenStore.snapshot() != expectedCredentials
            ) {
                return@synchronized false
            }
            _state.value = if (session.mustChangePassword) {
                AuthState.MustChangePassword(session)
            } else {
                AuthState.SignedIn(session)
            }
            true
        }

    private fun clearCurrentCredentialsAndSignOut(
        credentials: CredentialSnapshot,
        generation: Long,
    ): Boolean = synchronized(credentialTransitionLock) {
        if (authGeneration.get() != generation) return@synchronized false
        if (!tokenStore.clearIfCurrent(credentials)) return@synchronized false
        authGeneration.incrementAndGet()
        _state.value = AuthState.SignedOut
        true
    }

    private suspend fun fetchSession(): UserSession? {
        val result = convex.query("users:sessionContext")
        val user = result.optJSONObject("user") ?: return null
        if (user.length() == 0) return null
        return UserSession(
            userId = user.optString("_id"),
            email = user.optString("email"),
            name = user.optString("name").ifBlank { user.optString("email") },
            role = user.optString("role", "user"),
            status = user.optString("status", "active"),
            mustChangePassword = user.optBoolean("mustChangePassword", false),
            departmentName = result.optJSONObject("department")?.optString("name"),
            positionName = result.optJSONObject("position")?.optString("name"),
            positionLevel = result.optJSONObject("position")?.optInt("level"),
        )
    }

    private fun ConvexException.invalidatesCredentials(): Boolean =
        code == "NO_SESSION" ||
            code.contains("Unauthenticated", ignoreCase = true) ||
            code.contains("Authentication", ignoreCase = true) ||
            code == "UNAUTHENTICATED"

    private fun rollbackAccountSwitch(
        generation: Long,
        credentialRevision: Long,
        issuedCredentials: CredentialSnapshot?,
        previousCredentials: CredentialSnapshot,
        previousState: AuthState,
    ) = synchronized(credentialTransitionLock) {
        if (authGeneration.get() != generation) return@synchronized

        val restored = if (issuedCredentials == null) {
            tokenStore.snapshot()?.hasSameTokens(previousCredentials) == true
        } else {
            tokenStore.replaceIfCurrent(
                issuedCredentials,
                previousCredentials.accessToken,
                previousCredentials.refreshToken,
            )
        }
        if (restored) {
            _state.value = previousState
        } else if (tokenStore.clearIfRevision(credentialRevision)) {
            authGeneration.incrementAndGet()
            _state.value = AuthState.SignedOut
        }
    }

    private fun AuthState.isAuthenticated(): Boolean =
        this is AuthState.SignedIn || this is AuthState.MustChangePassword

    private fun CredentialSnapshot.hasSameTokens(other: CredentialSnapshot): Boolean =
        accessToken == other.accessToken && refreshToken == other.refreshToken

    private data class SignInAttempt(
        val previousCredentials: CredentialSnapshot?,
        val previousState: AuthState,
        val generation: Long,
        val credentialRevision: Long,
    )
}

File: ios-app/LvtCrm/Data/Auth/AuthRepository.swift
import Foundation

@MainActor
final class AuthRepository: ObservableObject {
    @Published private(set) var state: AuthState = .loading

    private let tokenStore: CredentialStore
    private let convex: ConvexHttpClient
    private let beforeSignOut: ((String) async -> Void)?
    private var authGeneration: Int64 = 0
    private let lock = NSLock()

    private let afterAuthenticated: (() async -> Void)?

    init(
        tokenStore: CredentialStore,
        convex: ConvexHttpClient,
        beforeSignOut: ((String) async -> Void)? = nil,
        afterAuthenticated: (() async -> Void)? = nil
    ) {
        self.tokenStore = tokenStore
        self.convex = convex
        self.beforeSignOut = beforeSignOut
        self.afterAuthenticated = afterAuthenticated
        Task { await restoreSession() }
    }

    func restoreSession() async {
        let generation = authGeneration
        guard let credentials = tokenStore.snapshot() else {
            if authGeneration == generation { state = .signedOut }
            return
        }
        do {
            if let session = try await fetchSession() {
                _ = publish(session, generation: generation, expectedCredentials: credentials)
                await afterAuthenticated?()
            } else {
                _ = clearCurrentCredentialsAndSignOut(credentials, generation: generation)
            }
        } catch let error as ConvexException
            where error.code.localizedCaseInsensitiveContains("Unauthenticated")
                || error.code.localizedCaseInsensitiveContains("Authentication") {
            _ = clearCurrentCredentialsAndSignOut(credentials, generation: generation)
        } catch {
            if authGeneration == generation, tokenStore.snapshot() == credentials {
                state = .signedOut
            }
        }
    }

    func signIn(email: String, password: String) async -> Result<UserSession, Error> {
        let normalized = email.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !normalized.isEmpty, !password.isEmpty else {
            return .failure(ConvexException(code: "EMAIL_PASSWORD_REQUIRED", message: "Nhập email và mật khẩu."))
        }

        let previousCredentials = tokenStore.snapshot()
        let previousState = state
        authGeneration += 1
        let generation = authGeneration
        let credentialRevision = tokenStore.invalidatePendingWrites()
        if previousCredentials != nil, previousState.isAuthenticated {
            state = .loading
        }

        var credentialsIssued = false
        var issuedCredentials: CredentialSnapshot?

        do {
            let result = try await convex.action(
                "auth:signIn",
                args: [
                    "provider": "password",
                    "params": [
                        "email": normalized,
                        "password": password,
                        "flow": "signIn",
                    ] as [String: Any],
                ],
                authenticated: false
            )
            guard let tokens = result["tokens"] as? [String: Any],
                  let access = tokens["token"] as? String,
                  let refresh = tokens["refreshToken"] as? String
            else {
                throw ConvexException(code: "NO_TOKENS", message: "Đăng nhập không trả về token.")
            }

            credentialsIssued = authGeneration == generation
                && tokenStore.saveIfRevision(credentialRevision, accessToken: access, refreshToken: refresh)
            guard credentialsIssued else {
                return .failure(ConvexException(code: "SIGN_IN_SUPERSEDED"))
            }
            issuedCredentials = CredentialSnapshot(
                accessToken: access,
                refreshToken: refresh,
                revision: credentialRevision
            )

            guard let session = try await fetchSession() else {
                throw ConvexException(code: "NO_SESSION", message: "Không tải được phiên đăng nhập.")
            }
            guard publish(session, generation: generation) else {
                return .failure(ConvexException(code: "SIGN_IN_SUPERSEDED"))
            }
            await afterAuthenticated?()
            return .success(session)
        } catch {
            if let previousCredentials, previousState.isAuthenticated {
                rollbackAccountSwitch(
                    generation: generation,
                    credentialRevision: credentialRevision,
                    issuedCredentials: issuedCredentials,
                    previousCredentials: previousCredentials,
                    previousState: previousState
                )
            } else if credentialsIssued, let issued = issuedCredentials, invalidatesCredentials(error) {
                _ = clearCurrentCredentialsAndSignOut(issued, generation: generation)
            }
            return .failure(error)
        }
    }

    func requestPasswordReset(email: String) async -> Result<Void, Error> {
        let normalized = email.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !normalized.isEmpty else {
            return .failure(ConvexException(code: "EMAIL_REQUIRED", message: "Nhập email."))
        }
        do {
            _ = try await convex.action(
                "users:requestPasswordReset",
                args: ["email": normalized],
                authenticated: false
            )
            return .success(())
        } catch {
            return .failure(error)
        }
    }

    func changePassword(newPassword: String) async -> Result<Void, Error> {
        let generation = authGeneration
        guard newPassword.count >= 8 else {
            return .failure(ConvexException(code: "PASSWORD_TOO_SHORT"))
        }
        do {
            _ = try await convex.action("users:changeOwnPassword", args: ["newPassword": newPassword])
        } catch {
            return .failure(error)
        }
        do {
            if let session = try await fetchSession() {
                _ = publish(session, generation: generation)
            } else if authGeneration == generation {
                authGeneration += 1
                tokenStore.clear()
                state = .signedOut
            }
            return .success(())
        } catch {
            return .success(())
        }
    }

    func signOut() {
        let accessToken: String?
        authGeneration += 1
        accessToken = tokenStore.accessToken
        tokenStore.clear()
        state = .signedOut
        guard let accessToken, !accessToken.isEmpty else { return }
        Task {
            if let beforeSignOut {
                await beforeSignOut(accessToken)
            }
            _ = try? await convex.actionWithToken("auth:signOut", args: [:], accessToken: accessToken)
        }
    }

    @discardableResult
    private func publish(
        _ session: UserSession,
        generation: Int64,
        expectedCredentials: CredentialSnapshot? = nil
    ) -> Bool {
        guard authGeneration == generation else { return false }
        if let expectedCredentials, tokenStore.snapshot() != expectedCredentials {
            return false
        }
        state = session.mustChangePassword ? .mustChangePassword(session) : .signedIn(session)
        return true
    }

    private func clearCurrentCredentialsAndSignOut(
        _ credentials: CredentialSnapshot,
        generation: Int64
    ) -> Bool {
        guard authGeneration == generation else { return false }
        guard tokenStore.clearIfCurrent(credentials) else { return false }
        authGeneration += 1
        state = .signedOut
        return true
    }

    private func fetchSession() async throws -> UserSession? {
        let result = try await convex.query("users:sessionContext")
        guard let user = result["user"] as? [String: Any], !user.isEmpty else { return nil }
        let department = result["department"] as? [String: Any]
        let position = result["position"] as? [String: Any]
        let email = (user["email"] as? String) ?? ""
        return UserSession(
            userId: (user["_id"] as? String) ?? "",
            email: email,
            name: ((user["name"] as? String)?.nilIfBlank) ?? email,
            role: (user["role"] as? String) ?? "user",
            status: (user["status"] as? String) ?? "active",
            mustChangePassword: (user["mustChangePassword"] as? Bool) ?? false,
            departmentName: department?["name"] as? String,
            positionName: position?["name"] as? String,
            positionLevel: position?["level"] as? Int
        )
    }

    private func invalidatesCredentials(_ error: Error) -> Bool {
        guard let error = error as? ConvexException else { return false }
        return error.code == "NO_SESSION"
            || error.code.localizedCaseInsensitiveContains("Unauthenticated")
            || error.code.localizedCaseInsensitiveContains("Authentication")
            || error.code == "UNAUTHENTICATED"
    }

    private func rollbackAccountSwitch(
        generation: Int64,
        credentialRevision: Int64,
        issuedCredentials: CredentialSnapshot?,
        previousCredentials: CredentialSnapshot,
        previousState: AuthState
    ) {
        guard authGeneration == generation else { return }
        let restored: Bool
        if issuedCredentials == nil {
            restored = tokenStore.snapshot()?.hasSameTokens(as: previousCredentials) == true
        } else if let issued = issuedCredentials {
            restored = tokenStore.replaceIfCurrent(
                issued,
                accessToken: previousCredentials.accessToken,
                refreshToken: previousCredentials.refreshToken
            )
        } else {
            restored = false
        }
        if restored {
            state = previousState
        } else if tokenStore.clearIfRevision(credentialRevision) {
            authGeneration += 1
            state = .signedOut
        }
    }
}

private extension CredentialSnapshot {
    func hasSameTokens(as other: CredentialSnapshot) -> Bool {
        accessToken == other.accessToken && refreshToken == other.refreshToken
    }
}

private extension String {
    var nilIfBlank: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}
